overlord/state/taskrunner.go

```diff
@@ -144,47 +187,66 @@ func (r *TaskRunner) Ensure() {
 	r.state.Lock()
 	defer r.state.Unlock()
 
-	for _, chg := range r.state.Changes() {
-		if chg.Status() == DoneStatus {
+	for _, t := range r.state.Tasks() {
+		handlers, ok := r.handlers[t.Kind()]
+		if !ok {
+			// Handled by a different runner instance.
 			continue
 		}
-		logger.Debugf("Working on change %s: %s", chg.ID(), chg.summary)
-
-		tasks := chg.Tasks()
-		for _, t := range tasks {
-			// done or error are final, nothing to do
-			// TODO: actually for error progate to halted and their waited
-			status := t.Status()
-			if status == DoneStatus || status == ErrorStatus {
-				continue
-			}
 
-			// No handler for the given kind of task,
-			// this means another taskrunner is going
-			// to handle this task.
-			if _, ok := r.handlers[t.Kind()]; !ok {
-				continue
-			}
+		tb := r.tombs[t.ID()]
 
-			// we look at the Tomb instead of Status because
-			// a task can be in DoStatus even when it
-			// is not started yet (like when the daemon
-			// process restarts)
-			if _, ok := r.tombs[t.ID()]; ok {
+		if t.Status() == AbortStatus {
+			if tb != nil {
+				tb.Kill(nil)
 				continue
 			}
+			r.swapAbort(t)
+		}
 
-			// check if there is anything we need to wait for
-			if mustWait(t) {
-				continue
+		if tb != nil {
+			// Already being handled.
+			continue
+		}
+
+		status := t.Status()
+		if status.Ready() {
+			continue
+		}
+		if status == UndoStatus && handlers.undo == nil {
+			// Cannot undo. Revert to done status.
+			t.SetStatus(DoneStatus)
```

Comment:
Ready, exactly.. we cannot undo something that is not undoable.. so we leave these alone. The tests verify that logic.